### PR TITLE
Attachment enable setting

### LINKED ISSUE #848: Configuration Option Allow/Disallow Attachments
**Is your feature request related to a problem? Please describe.**
Attachments on tickets are a wonderful addition to the ticketing system, however, they also have maintenance overhead and security risks. Some companies do not allow attachments on emails or ticket systems. In a simpler installation of the ticketing system, having the option to disable attachments can be beneficial from both maintenance and security standpoints. An alternative to save a file in the database or in the ticket systems is to allow the capability to include a link to a file location on a network.

**Describe the solution you'd like**
Create a configuration option HELPDESK_ALLOW_ATTACHMENTS that is set in the Django Project settings.py file.
If set to false it would remove the option from the menus and not allow the actions to occur. Cases where the option was allowed and files exist would need to be taken into account.

**Describe alternatives you've considered**
1) Monkey patch the Attachments class save method in the helpdesk models.py to just return. This is effective at disallowing attachments, but provides no lasting solution and can perplex users trying to attach files.
2) Have this option and/or alternative to use a link to a file location provided on a network.

### PR DESCRIPTION
**Pull request by jorge-leon**
Fixes #848. See also discussion  #1188

### CODE REVIEW

File: helpdesk/settings.py [hunk 1]
```diff
@@ -56,6 +56,11 @@
                                          'HELPDESK_STAFF_VIEW_PROTECTOR',
                                          lambda _: None)
 
+# Enable ticket and Email attachments
+HELPDESK_ENABLE_ATTACHMENTS = getattr(settings,
+                                      'HELPDESK_ENABLE_ATTACHMENTS',
+                                      False)
+
```
Comment (uhurusurfa) [inline helpdesk/settings.py:63]: Unless there is a strong argument to changing a default, I will always err on the side of remaining backwards compatible.
In this case I can see the argument that there is a security risk for new users who do not want attachments enabled but that is still an installation decision new users should make and as such, as long as we clearly document it in the install instructions, I do not see that it is necessary to make the default "False". 
This means that long time existing users will not be surprised by a change they did not expect when they upgrade and find they have lost a chunk of data for their tickets before they realise what the problem is.
I am open to discussion on this if there are strong opinions for making it default to False.

**Review by uhurusurfa — CHANGES_REQUESTED** [review]

File: helpdesk/settings.py (hunk 1 above)
Comment (jorge-leon) [reply helpdesk/settings.py:63]: No strong argument at all.

I'm always happy to find, that makers of open source package 'X' have my back covered when I start dabbling with 'X', but I also understand, that breaking software with an update will always cause confusion and anger.

I propose to
- Expand the security topic in the documentation
- Set safety default in a future major upgrade with due announcements.

For now I will update the PR to default to enabled attachments.

**Comment by jorge-leon** [comment]
Testing fails with `Queue processing failed: test2 -- Error Q2`, I guess when processing a (mocked?) Email queue.

Can you give pointers how to test/mock Email on the development system in order to look into this?  The pytest output does not give clues and I'm unfamiliar with pytesting.

**Comment by uhurusurfa** [comment]
> Testing fails with `Queue processing failed: test2 -- Error Q2`, I guess when processing a (mocked?) Email queue.
> 
> Can you give pointers how to test/mock Email on the development system in order to look into this? The pytest output does not give clues and I'm unfamiliar with pytesting.

You PR is passing right now but generally you can do this from the Helpdesk project base folder in a shell(I am using *nix notation so use backslash if you are Windows):

1. Run all test by executing quicktest.py ./quicktest.py (or: python3 quicktest.py if python3 is not the default python for your system)
2. Run a specific test use the format: ./quicktest.py <fully qualified path to a test clas or method within a test class>
           eg. ./quicktest.py helpdesk.tests.test_get_email.GetEmailCommonTests.test_email_with_txt_as_attachment_with_simple_alternative_message

NOTE: You must be in a virtualenv shell when running the tests so you can do this to activate a virtualenv from the Helpdesk project base folder assuming you hve set up a virtualenv per the "Developer Environment" instructions in README.rst:
`.  ./.venv/bin/activate`

Alternatively you can use the debugger in whatever

**Review by uhurusurfa — APPROVED** [review]
🚀  Looks good to me. @Benbb96  do you have any firther input for this PR?

**Comment by jorge-leon** [comment]
...
>     1. Run all test by executing quicktest.py ./quicktest.py (or: python3 quicktest.py if python3 is not the default python for your system)
> 
... 
> Alternatively you can use the debugger in whatever

Thank you for the instructions.  It looks like Q2 and Q4 are expected to throw an exception during test.

File: helpdesk/templates/helpdesk/ticket.html [hunk 2]
```diff
@@ -197,7 +199,9 @@ <h5>{{ checklist }}</h5>
             </div>
         {% endif %}
 
+        {% if helpdesk_settings.HELPDESK_ENABLE_ATTACHMENTS %}
         <p id='ShowFileUploadPara'><button type="button" class="btn btn-warning btn-sm" id='ShowFileUpload'>{% trans "Attach File(s) &raquo;" %}</button></p>
```
Comment (Benbb96) [inline helpdesk/templates/helpdesk/ticket.html:203]: There are also related parts below in JS (line 278, and 307 to the end) that can be conditionally displayed.

**Review by Benbb96 — COMMENTED** [review]
One minor thing, either way it looks good to me too. Thanks for the contribution @jorge-leon !

File: helpdesk/templates/helpdesk/ticket.html (hunk 2 above)
Comment (jorge-leon) [reply helpdesk/templates/helpdesk/ticket.html:203]: I hoped to get away without touching JS. But here we are.

Can you confirm that the following lines are the ones related to file handling?
- 278-282: `$("#ShowFileUpload")`...
- 307-334: `$('.add_file_fields_wrap)?`... and `$(document).on('change', ':file'`...

Comment (Benbb96) [reply helpdesk/templates/helpdesk/ticket.html:203]: Yes I confirm. These lines shouldn't trigger errors in console but it's better not to have them in DOM if attachments are disabled.

File: docs/settings.rst [hunk 3]
```diff
@@ -51,16 +51,30 @@ These settings can be used to change who can access the helpdesk.
...
+Settings related to attachments:
 
-- **HELPDESK_ANON_ACCESS_RAISES_404** If True, redirects user to a 404 page when attempting to reach ticket pages while not logged in, rather than redirecting to a login screen.
+- **HELPDESK_ENABLE_ATTACHMENTS** If set to ``True``, files can be
+  attached to tickets and followups, and emails are searched for
+  attachments which are then attached to the ticket.  Also enables the
+  ``HELPDESK_ALWAYS_SAVE_INCOMING_EMAIL_MESSAGE`` setting.
 
-  **Default:** ``HELPDESK_ANON_ACCESS_RAISES_404 = False``
+  **Caution**: Set this to False, unless you have secured access to
```
Comment (timthelion) [inline docs/settings.rst:65]: This is false FUD. Unless you have a non-standard Media setup that allows file listing, attachment names are protected by the ticket's secret key and cannot be guessed:

https://github.com/django-helpdesk/django-helpdesk/blob/c77bb30035bfc32d00dc6aab635b38f18e6538c4/helpdesk/models.py#L1218

**Comment by timthelion** [comment]
https://github.com/django-helpdesk/django-helpdesk/commit/caaa0e515844f61b2c47cb29fb320dfcee0d45a5

**Review by uhurusurfa — APPROVED** [review]
👍 Looks good to me.

**Comment by timthelion** [comment]
It still contains the mostly false warning:

> 
> **Caution**: Set this to False, unless you have secured access to the uploaded files. Otherwise anyone on the Internet will be able to download your ticket attachments.
> Attachments are enabled by default for backwards compatibility.
> 

Why include lies and FUD in our docs?

A better text would be:

> 
> **Caution**: With most Django Media back-ends attachments can be downloaded by anyone who knows the unique and unguessable URL. If you want greater security choose a Media back-end which is secured or disable attachments by setting this setting to False.

**Comment by jorge-leon** [comment]
I like the wording provided by @timthelion, it makes things clear and shows how to act on.

If all agree and if not done before I'll be happy to 'sneak' it in with some future PR.

**Pull request merged**